fvm/src/blockstore/buffered.rs

```diff
@@ -62,86 +56,63 @@ where
-        if val <= u8::MAX as u16 {
-            return Err(anyhow!(
-                "cbor input was not canonical (lval 25 with value <= MaxUint8)"
-            ));
-        }
-        Ok((maj, val as usize))
-    } else if low == 26 {
-        br.read_exact(&mut scratch[..4])?;
-        let val = BigEndian::read_u32(&scratch[..4]);
-        if val <= u16::MAX as u32 {
-            return Err(anyhow!(
-                "cbor input was not canonical (lval 26 with value <= MaxUint16)"
-            ));
-        }
-        Ok((maj, val as usize))
-    } else if low == 27 {
-        br.read_exact(&mut scratch[..8])?;
-        let val = BigEndian::read_u64(&scratch[..8]);
-        if val <= u32::MAX as u64 {
-            return Err(anyhow!(
-                "cbor input was not canonical (lval 27 with value <= MaxUint32)"
-            ));
-        }
-        Ok((maj, val as usize))
-    } else {
-        Err(anyhow!("invalid header cbor_read_header_buf"))
-    }
+    let val = match low {
+        ..=23 => low.into(),
+        24 => read_fixed::<1>(br)?[0].into(),
+        25 => u16::from_be_bytes(read_fixed(br)?).into(),
+        26 => u32::from_be_bytes(read_fixed(br)?).into(),
+        27 => u64::from_be_bytes(read_fixed(br)?),
+        _ => return Err(anyhow!("invalid header cbor_read_header_buf")),
+    };
+    Ok((maj, val))
 }
 
 /// Given a CBOR serialized IPLD buffer, read through all of it and return all the Links.
 /// This function is useful because it is quite a bit more fast than doing this recursively on a
 /// deserialized IPLD object.
-fn scan_for_links<B: Read + Seek, F>(buf: &mut B, mut callback: F) -> Result<()>
-where
-    F: FnMut(Cid) -> anyhow::Result<()>,
-{
-    let mut scratch: [u8; 100] = [0; 100];
+fn scan_for_links(mut buf: &[u8], out: &mut Vec<Cid>) -> Result<()> {
     let mut remaining = 1;
     while remaining > 0 {
-        let (maj, extra) = cbor_read_header_buf(buf, &mut scratch)?;
+        let (maj, extra) = cbor_read_header_buf(&mut buf)?;
         match maj {
             // MajUnsignedInt, MajNegativeInt, MajOther
             0 | 1 | 7 => {}
             // MajByteString, MajTextString
             2 | 3 => {
-                buf.seek(std::io::SeekFrom::Current(extra as i64))?;
+                if extra > buf.len() as u64 {
+                    return Err(anyhow!("unexpected end of cbor stream"));
+                }
+                buf = &buf[extra as usize..];
             }
             // MajTag
             6 => {
                 // Check if the tag refers to a CID
                 if extra == 42 {
-                    let (maj, extra) = cbor_read_header_buf(buf, &mut scratch)?;
+                    let (maj, extra) = cbor_read_header_buf(&mut buf)?;
                     // The actual CID is expected to be a byte string
                     if maj != 2 {
                         return Err(anyhow!("expected cbor type byte string in input"));
                     }
-                    if extra > 100 {
```

Comment:
This was unnecessary. If the CID is invalid and/or too big, `Cid::try_from` will catch that.